### PR TITLE
feat: add mdbook preprocessing for forc-publish to flatten include directives

### LINKED ISSUE #7263: forc-publish: mdbook preprocessor
We need to do a pre-process pass over the readme to handle include directives. This is generally used with more complex project structures and needed by sway-libs:

Basically if there is a readme.md file with the following content:

```
{{#include ../../docs/book/src/admin/index.md}}
```

we should go and read that file and flatten. This should happen in the correct dependency order.

### PR DESCRIPTION
**Pull request by kayagokalp**
## Description
closes #7263 

This pull request introduces functionality to preprocess Markdown files by handling `{{#include}}` directives, enabling the inclusion of content from other files. It also integrates this functionality into the `forc-publish` plugin to process `README.md` files during tarball creation. The most important changes include the addition of the Markdown preprocessing module, updates to error handling, and modifications to the tarball creation process.

### Markdown Preprocessing Functionality:
* [`forc-plugins/forc-publish/src/md_pre_process/mod.rs`](diffhunk://#diff-1d065227f3eea45f66a3189882f220de977df2e1b7d1fb2de36d2b9baeb39a7aR1-R293): Added a new module to preprocess Markdown files, including features like dependency graph construction, topological sorting, and content flattening for `{{#include}}` directives. Includes comprehensive tests for edge cases like cycles and missing includes.
* [`forc-plugins/forc-publish/src/md_pre_process/error.rs`](diffhunk://#diff-2cb6edf1287489eb7edd794db5e8b727907a28755871f656cfce877aadd886c2R1-R25): Introduced a dedicated error type `MDPreProcessError` to handle issues like I/O errors, missing include files, and cycles in Markdown dependencies.

### Integration with `forc-publish` Plugin:
* [`forc-plugins/forc-publish/src/lib.rs`](diffhunk://#diff-8b5115a5299062687979272a86a86b573a688194c06c044174da61798ef41574R4): Registered the new `md_pre_process` module in the plugin library.
* [`forc-plugins/forc-publish/src/tarball.rs`](diffhunk://#diff-e3f20cadab3a1cf99fc967e0adc12e908f435a185e1529df9b152388927adf45R29-R31): Updated the tarball creation process to preprocess `README.md` files using the `flatten_markdown` function. Added a warning mechanism to log errors without failing the publish process. [[1]](diffhunk://#diff-e3f20cadab3a1cf99fc967e0adc12e908f435a185e1529df9b152388927adf45R29-R31) [[2]](diffhunk://#diff-e3f20cadab3a1cf99fc967e0adc12e908f435a185e1529df9b152388927adf45R44-R60)

### Error Handling Enhancements:
* [`forc-plugins/forc-publish/src/error.rs`](diffhunk://#diff-d09f44c4690282b827488db75a21e561e1e6fa72d549904db3f5fc4313f4819cR34-R36): Extended the `Error` enum to include `MDPreProcessError` for better error propagation during Markdown preprocessing.

### Dependency Updates:
* [`forc-plugins/forc-publish/Cargo.toml`](diffhunk://#diff-2bd34a7b1fa0f1850ad25c6e080d410b2df98c16b6cd75445a138d04e617311aR29): Added `regex` as a workspace dependency to support Markdown parsing.

### CODE REVIEW

File: forc-plugins/forc-publish/src/md_pre_process/mod.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        let re = Regex::new(r"\{\{#include\s+([^\}]+)\}\}")?;
+
+        let includes = re
+            .captures_iter(&content)
+            .filter_map(|caps| {
+                let inc_rel = caps[1].trim();
+                let inc_path = dir.join(inc_rel);
+                inc_path.canonicalize().ok()
+            })
+            .collect();
```
Comment (zees-dev) [inline forc-plugins/forc-publish/src/md_pre_process/mod.rs:34]: nit: not sure if you want to explicitly catch errors for canonicalization
```suggestion
        let includes = re
            .captures_iter(&content)
            .map(|caps| {
                let inc_rel = caps[1].trim();
                let inc_path = dir.join(inc_rel);
                inc_path
                    .canonicalize()
                    .map_err(|_| MDPreProcessError::Canonicalize(inc_path))
            })
            .collect::<Result<HashSet<_>, _>>()?;
```

File: forc-plugins/forc-publish/src/md_pre_process/mod.rs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+    #[test]
+    fn test_flatten_single_file_no_includes() -> Result<(), MDPreProcessError> {
+        let dir = tempdir()?;
+        let readme = dir.path().join("README.md");
+        write_file(&readme, "# Title\n\nHello world!")?;
+
+        let result = flatten_markdown(&readme)?;
+        assert!(result.contains("Hello world!"));
+        assert!(result.contains("# Title"));
```
Comment (zees-dev) [inline forc-plugins/forc-publish/src/md_pre_process/mod.rs]: nit: better to assert exact content of result (since it isnt much anyway); same for tests below. Makes it easier to compare before and after markdown operation.

**Review by zees-dev — DISMISSED** [review]
few small nits; LGTM 👍

File: forc-plugins/forc-publish/src/md_pre_process/mod.rs (hunk 2 above)
Comment (kayagokalp) [reply forc-plugins/forc-publish/src/md_pre_process/mod.rs]: addressed in [a4c6451](https://github.com/FuelLabs/sway/pull/7264/commits/a4c64516bb2cc0a08ad9b264894a9f81183b0735)
